File: tools/Changelog/Changelog.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#      Copyright (C) 2007-2010 Team XBMC
#      http://www.xbmc.org
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with XBMC; see the file COPYING.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html

import re, os, sys

def usage ():
  print " Changelog.py [option]"
  print "  -h       : display this message"
  print "  -r <REV> : get log messages up until REV"
  print "  -d <DIR> : place Changelog.txt in DIR"
  sys.exit()


header =  "*************************************************************************************************************\n" \
          "*************************************************************************************************************\n" \
          "                                     XBMC CHANGELOG\n" \
          "*************************************************************************************************************\n" \
          "*************************************************************************************************************" \
          "\nDate        Rev   Message\n" \
          "=============================================================================================================\n"

xmlre = re.compile("^<logentry.*?revision=\"([0-9]{4,})\".*?<date>([0-9]{4}-[0-9]{2}-[0-9]{2})T.*?<msg>(.*?)</msg>.*?</logentry>$", re.MULTILINE | re.DOTALL)
txtre = re.compile("([0-9]{4}-[0-9]{2}-[0-9]{2})  ([0-9]{4,5}) {1,2}(.*)")

old = None
lastrev = 8638
rev = 1000000
dir = "."

nargs = len(sys.argv)
args = sys.argv

i = 1
while i < nargs:
  if args[i] == "-r":
    i += 1
    try:
      rev = int(sys.argv[i])
    except:
      rev=1000000
  elif args[i] == "-d":
    i+=1
    dir = args[i].replace(' ', '\ ')
  elif args[i] == "-h" or args[i] == "--help":
    usage()
  i+=1

# print dir
# print rev

try:
  old = open("%s/Changelog.txt" % (dir))
except:
  old = None

if old != None:
  olddoc = old.read()
  old.close()
  oldmsgs = txtre.findall(olddoc)
  del olddoc
  if len(oldmsgs) > 0:
    lastrev = int(oldmsgs[0][1])
try:
  output = open("%s/Changelog.txt" % (dir),"w")
except:
  print "Can't open %s/Changelog.txt for writing." % (dir)
  sys.exit()

output.write(header)

if rev <= lastrev:
  for msg in oldmsgs:
    if int(msg[1]) <= rev:
      s = "%-11.11s %-5.5s %s\n" % (msg[0], msg[1], msg[2])
      output.write(s)
  sys.exit()

svncmd = "svn log --xml -r %s:HEAD" % (lastrev)
newlog = os.popen(svncmd)

newlogdoc = newlog.read()
newlog.close()

newmsgs = xmlre.findall(newlogdoc)

newmsgs.reverse()

for msg in newmsgs:
  s = "%-11.11s %-5.5s %s\n" % (msg[1], msg[0], msg[2].replace('\n',''))
  output.write(s)

skip = 0
if old != None:
  for msg in oldmsgs:
    if skip == 1:
      s = "%-11.11s %-5.5s %s\n" % (msg[0], msg[1], msg[2])
      output.write(s)
    else:
      skip = 1

output.close()

File: XBMC/tools/Changelog/Changelog.cpp
/*
 *      Copyright (C) 2005-2008 Team XBMC
 *      http://www.xbmc.org
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with XBMC; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */

// Changelog.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"

#include "tinyxml.h"
#include "../../xbmc/utils/RegExp.h"

#ifdef _LINUX
#define _tmain main
#define _TCHAR char
#endif

const char header[] = "*************************************************************************************************************\r\n"
                      "*************************************************************************************************************\r\n"
                      "                                     XBMC CHANGELOG\r\n"
                      "*************************************************************************************************************\r\n"
                      "*************************************************************************************************************\r\n"
                      "\r\n"
                      "Date        Rev   Message\r\n"
                      "=============================================================================================================\r\n";

const char filter[][100] = {"[- ]*[0-9]+-[0-9]+-[0-9]+ *",
                             "\\*\\*\\* empty log message \\*\\*\\*",
                             "no message" };

std::string FilterMessage(std::string message)
{
  std::string filteredMessage = message;
  CRegExp reg;
  for (int i = 0; i < sizeof(filter) / 100; i++)
  {
    reg.RegComp(filter[i]);
    int findStart = reg.RegFind(message.c_str());
    while (findStart >= 0)
    {
      filteredMessage = message.substr(0, findStart);
      filteredMessage.append(message.substr(findStart + reg.GetFindLen(), message.length()));
      message = filteredMessage;
      findStart = reg.RegFind(message.c_str());
    }
  }
  return filteredMessage;
}

int _tmain(int argc, _TCHAR* argv[])
{
  std::string input = "svn_log.xml";
  std::string output = "Changelog.txt";
  int limit = 0;

  if (argc < 2)
  {
    // output help information
    printf("usage:\n");
    printf("\n");
    printf("  Changelog input <output> <limit>\n");
    printf("\n");
    printf("  input    : input .xml file generated from SVN (using svn log --xml)\n");
    printf("             DOWNLOAD to download direct from XBMC SVN\n");
    printf("  <output> : output .txt file for the changelog (defaults to Changelog.txt)\n");
    printf("  <limit>  : the number of log entries for svn to fetch. (defaults to no limit)");
    printf("\n");
    return 0;
  }
  input = argv[1];
  if (argc > 2)
    output = argv[2];
  FILE *file = fopen(output.c_str(), "wb");
  if (!file)
    return 1;
  fprintf(file, header);
  if (input.compare("download") == 0)
  {
    if(argc > 3)
      limit = atoi(argv[3]);
    // download our input file
    std::string command = "svn log -r 'HEAD':8638 ";
    if (limit > 0)
    {
      command += "--limit ";
      command += argv[3]; // the limit as a string
      command += " ";
    }
#ifndef _LINUX
    command += "--xml https://xbmc.svn.sourceforge.net/svnroot/xbmc/trunk/XBMC > svn_log.xml";
#else
    command += "--xml https://xbmc.svn.sourceforge.net/svnroot/xbmc/branches/linuxport/XBMC > svn_log.xml";
#endif
    printf("Downloading changelog from SVN - this will take some time (around 1MB to download with no limit)\n");
    system(command.c_str());
    input = "svn_log.xml";
    printf("Downloading done - processing\n");
  }
  TiXmlDocument doc;
  if (!doc.LoadFile(input.c_str()))
  {
    return 1;
  }

  TiXmlElement *root = doc.RootElement();
  if (!root) return 1;

  TiXmlElement *logitem = root->FirstChildElement("logentry");
  while (logitem)
  {
    int revision;
    logitem->Attribute("revision", &revision);
    TiXmlNode *date = logitem->FirstChild("date");
    std::string dateString;
    if (date && date->FirstChild())
      dateString = date->FirstChild()->Value();
    TiXmlNode *msg = logitem->FirstChild("msg");
    if (msg && msg->FirstChild())
    {
      // filter the message a bit
      std::string message = FilterMessage(msg->FirstChild()->Value());
      if (message.size())
        fprintf(file, "%s  %4i  %s\r\n", dateString.substr(0,10).c_str(), revision, message.c_str());
      else
        int breakhere = 1;
    }
    logitem = logitem->NextSiblingElement("logentry");
  }
  fclose(file);
  printf("Changelog saved as: %s\n", output.c_str());
  return 0;
}
